### DISCUSSION TITLE
Incorrect key detection on Portuguese Thinkpad keyboard layout

### DISCUSSION THREAD

**Discussion by luissimas** [ghostty-org/ghostty#4131] [Feature Requests, Ideas] [answered]
Hello folks! I've been using Ghostty for the past couple of days and it's been a very nice experience, thanks a lot!

I'm using a Thinkpad T480 with a Portuguese (Brasil, IBM/Lenovo ThinkPad) keyboard layout and I noticed that the `/` key is not being detected as expected in this particular keyboard layout. Switching to a standard US keyboard produces the correct behavior. I compared the behavior with Kitty and Alacritty and found that Ghostty differs from both of them.

Here's the output of running `kitten show-key -m kitty` in different terminals and pressing the `/` key. Ghostty is on the left, while Alacritty is on the top right and Kitty is on the bottom right:

![image](https://github.com/user-attachments/assets/588f4a2d-c51b-4ca6-9382-2017b93e8287)

From what I understand, Ghostty detects the `/` keypress in this layout as a `ctrl+/` keypress.

I'm using GNOME 47.1 on Fedora 41 Workstation. Here's the output of `ghostty +version`:

```
❯ ghostty +version
Ghostty 1.0.1-main+d512f560

Version
  - version: 1.0.1-main+d512f560
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.gtk
  - font engine: font.main.Backend.fontconfig_freetype
  - renderer   : renderer.OpenGL
  - libxev     : main.Backend.io_uring
  - GTK version:
    build      : 4.16.5
    runtime    : 4.16.5
  - libadwaita : enabled
    build      : 1.6.2
    runtime    : 1.6.2
  - libX11     : enabled
  ```

**Comment by rhodes-b** [comment]
not sure if portugese remaps the `/` key but can you try the `physical` modifier for the keybind you are attemping to use
you can see a note about that here
https://ghostty.org/docs/config/reference#keybind

**Reply by luissimas** [reply]
Thanks for the tip! 

I'm not sure if I didn't bind it correctly, but I tried the following binding and it seemed to have no effect:

```
keybind = physical:slash=text:/
```

**Reply by rhodes-b** [reply]
you can use the terminal inspector to see what input ghostty sees the key presses as (I pressed backslash in this case)
![image](https://github.com/user-attachments/assets/3becb30a-9cea-4a24-8d53-f73db2d601da)

**Reply by rhodes-b** [reply]
~~Looks like it needs to escape backslash not sure if that is intented with the `text` field or not then it seems to work to send backslash~~
~~`keybind = physical:slash=text:\\`~~

edit: this is irrelivant

**Reply by luissimas** [reply]
The output of the inspector was the same as using `kitten show-key -m kitty`. It seems that the key is being identified as `ctrl+/` instead of `/`.

![image](https://github.com/user-attachments/assets/f6092d74-a107-470f-8395-3878d67c8f22)

I was able to get the key to work correctly by doing the following binding:

```
keybind = ctrl+/=text:/
```

I'd say it's a valid workaround (although now I cannot use the actual `ctrl+/` properly), but still does not seem to be the intended behavior. I'd be willing to investigate this further and submit a fix if it makes sense for the project. If that sounds good, I'd just like to get a few pointers on where to start looking for.

**Comment by luissimas** [comment]
I decided to launch Ghostty via the shell and reproduce the behavior to see if there's anything in the logs, and it seems that the following log line is printed everytime I press the `/` key:

```
info(io_handler): BELL
```

Doing a quick grep on the repository, it seems that this log originates on this line:

https://github.com/ghostty-org/ghostty/blob/d6e76858164d52cff460fedc61ddf2e560912d71/src/termio/stream_handler.zig#L330

When using an international US layout, pressing the `/` key does not produce any log output, so this may be related to the issue.

Here's the full log output of doing the following:

1. Open Ghostty via a shell
2. Press `/` three times
3. Close the Ghostty window

```
❯ ghostty
info: ghostty version=1.0.1-main+d512f560
info: ghostty build optimize=ReleaseFast
info: runtime=apprt.Runtime.gtk
info: font_backend=font.main.Backend.fontconfig_freetype
info: dependency harfbuzz=8.4.0
info: dependency fontconfig=21402
info: renderer=renderer.OpenGL
info: libxev backend=main.Backend.io_uring
info(os): setlocale from env result=en_US.UTF-8
info(gtk): GTK version build=4.16.5 runtime=4.16.5
info: reading configuration file path=/home/padawan/.config/ghostty/config
info(config): default shell source=env value=/bin/bash
info(gtk): libadwaita version build=1.6.2 runtime=1.6.2
info(config): default shell source=env value=/bin/bash
EGL API version 1.5 found
 - Vendor: Mesa Project
 - Version: 1.5
 - Device: /dev/dri/renderD128, 226 128
 - Client APIs: OpenGL OpenGL_ES
 - Extensions:
        EGL_ANDROID_blob_cache
        EGL_ANDROID_native_fence_sync
        EGL_EXT_buffer_age
        EGL_EXT_config_select_group
        EGL_EXT_create_context_robustness
        EGL_EXT_image_dma_buf_import
        EGL_EXT_image_dma_buf_import_modifiers
        EGL_EXT_pixel_format_float
        EGL_EXT_present_opaque
        EGL_EXT_query_reset_notification_strategy
        EGL_EXT_surface_compression
        EGL_EXT_swap_buffers_with_damage
        EGL_IMG_context_priority
        EGL_KHR_cl_event2
        EGL_KHR_config_attribs
        EGL_KHR_context_flush_control
        EGL_KHR_create_context
        EGL_KHR_create_context_no_error
        EGL_KHR_fence_sync
        EGL_KHR_get_all_proc_addresses
        EGL_KHR_gl_colorspace
        EGL_KHR_gl_renderbuffer_image
        EGL_KHR_gl_texture_2D_image
        EGL_KHR_gl_texture_3D_image
        EGL_KHR_gl_texture_cubemap_image
        EGL_KHR_image_base
        EGL_KHR_no_config_context
        EGL_KHR_reusable_sync
        EGL_KHR_surfaceless_context
        EGL_KHR_swap_buffers_with_damage
        EGL_KHR_wait_sync
        EGL_MESA_configless_context
        EGL_MESA_drm_image
        EGL_MESA_gl_interop
        EGL_MESA_image_dma_buf_export
        EGL_MESA_query_driver
        EGL_MESA_x11_native_visual_id
        EGL_WL_bind_wayland_display
        EGL_WL_create_wayland_buffer_from_image
 - Selected fbconfig: R8G8B8A8, depth 0, stencil 0
          high depth: R16G16B16A16 float, depth 0, stencil 0
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
Using OpenGL backend EGL
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
Creating EGL context version 3.3 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
info(grid): loaded OpenGL 4.6
info(font_shared_grid_set): font regular: IosevkaTerm NF
info(font_shared_grid_set): font bold: IosevkaTerm NF Bold
info(font_shared_grid_set): font italic: IosevkaTerm NF Italic
info(font_shared_grid_set): font bold_italic: IosevkaTerm NF Bold Italic
info(io_exec): found Ghostty resources dir: /home/padawan/.local/share/ghostty
info(io_exec): shell integration automatically injected shell=termio.shell_integration.Shell.bash
warning(gtk): unimplemented action=apprt.action.Action.Key.cell_size
warning(gtk): unimplemented action=apprt.action.Action.Key.size_limit
info(io_exec): started subcommand path=/bin/sh pid=172954
info(io_exec): subcommand cgroup=-
Creating EGL context version 3.2 (debug:no, forward:no, legacy:no, es:no)
OpenGL version: 4.6 (core)
GLSL version: 4.60
Max texture size: 16384
Enabled features (use GDK_GL_DISABLE env var to disable):
    debug: ✓
    unpack-subimage: ✓
    half-float: ✓
    sync: ✓
    base-instance: ✓
    buffer-storage: ✓
Using fractional scale 1 for EGL window (1000 600 => 1000 600)
info(grid): reallocating GPU buffer old=0 new=8
info(grid): reallocating GPU buffer old=8 new=38
info(io_handler): BELL
info(io_handler): BELL
info(io_handler): BELL
Destroying EGL context
Destroying EGL context
info(io_exec): read thread got quit signal
info(surface): surface closed addr=13e19600
info(sentry): sentry envelope does not contain crash, discarding
```

**Reply by 00-kat** [reply]
The bell is Ctrl+G; C escape `\a`, `\007` or `\x07`, 0x07 in hex.

Although sending Ctrl+G to the running application (which is what pressing it when at a shell does) wouldn't result in `info(io_handler): BELL`, so I'm not sure what's up.

**Comment by luissimas** [comment]
I'll post an update here since I'll probably not look further into this today. I've been playing detective with this one (and having quite a bit of fun while doing so) and I made some progress on the investigation.

# Summary

The `/` key does not work properly. It produces `ctrl+/` instead of just `/`. One interesting thing is that in the Thinkpad portuguese layout, the right control key is replaced by the `/` key, so this might be the root cause of the issue.

When I press the key, Ghostty detects it as the `BEL` C0 control code, which has the value `0x07` in hex.

## Looking at the messages in termio

Analyzing the `drainMailbox` function on `src/termio/Thread.zig`, I found that the following gets logged with the Thinkpad keyboard:

```text
debug(io_thread): mailbox message=termio.message.Message{ .write_small = termio.message.MessageData(u8,38).Small{ .data = { 31, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170 }, .len = 1 } }
debug(app): mailbox message=redraw_surface
info(stream): execute: c0terminal.ansi.C0.BEL
info(io_handler): BELL
```

While the following gets logged with the US international keyboard:

```text
debug(io_thread): mailbox message=termio.message.Message{ .write_small = termio.message.MessageData(u8,38).Small{ .data = { 47, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170 }, .len = 1 } }
debug(renderer_thread): mailbox message=renderer.message.Message{ .reset_cursor_blink = void }
```

We can se that in the international keyboard the first byte of the data is `47`, which is what Kitty and Alacritty show after `CSI 47` in the output of `kitten show-key -m kitty`.

I think the `CSI 47 ; 5 u` escape sequence show by Ghostty on `kitten show-key -m kitty` can be broke down as follows:

- `CSI` marks the start of the escape sequence
- `47` the keycode of the key
- `5` the modifier, which in this case is (incorrectly) detected as `Ctrl`
- `u` marks the end of the escape sequence

The fact that the logs are different at this stage indicate that the bug is at an earlier point, before the message is put into the thread's queue. The next step is to find out where this message is created.

## How the message is created

Eventually I figured out that the input events are handled in `src/Surface.zig` and encoded to keys in the `encodeKey` function, which calls the `encode` function in `src/input/KeyEncoder.zig`. Adding some logs there and reproducing the behavior gave me some more insights.

Using the Thinkpad keyboard:

```text
debug(key_encoder): encoding key buf={ 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170 } kitty_flags=terminal.kitty.key.Flags{ .disambiguate = false, .report_events = false, .report_alternates = false, .report_all = false, .report_associated = false } event=input.key.KeyEvent{ .action = input.key.Action.press, .key = input.key.Key.slash, .physical_key = input.key.Key.right_control, .mods = input.key.Mods{ .shift = false, .ctrl = true, .alt = false, .super = false, .caps_lock = false, .num_lock = false, .sides = input.key.Mods.side{ .shift = input.key.Mods.Side.left, .ctrl = input.key.Mods.Side.right, .alt = input.key.Mods.Side.left, .super = input.key.Mods.Side.left }, ._padding = 0 }, .consumed_mods = input.key.Mods{ .shift = false, .ctrl = false, .alt = false, .super = false, .caps_lock = false, .num_lock = false, .sides = input.key.Mods.side{ .shift = input.key.Mods.Side.left, .ctrl = input.key.Mods.Side.left, .alt = input.key.Mods.Side.left, .super = input.key.Mods.Side.left }, ._padding = 0 }, .composing = false, .utf8 = { 47 }, .unshifted_codepoint = 47 }
debug(surface): created write_req=termio.message.MessageData(u8,38){ .small = termio.message.MessageData(u8,38).Small{ .data = { 31, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170 }, .len = 1 } }
debug(io_thread): mailbox message=termio.message.Message{ .write_small = termio.message.MessageData(u8,38).Small{ .data = { 31, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170 }, .len = 1 } }
```

Using the US international keyboard:

```text
debug(key_encoder): encoding key buf={ 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170 } kitty_flags=terminal.kitty.key.Flags{ .disambiguate = false, .report_events = false, .report_alternates = false, .report_all = false, .report_associated = false } event=input.key.KeyEvent{ .action = input.key.Action.press, .key = input.key.Key.slash, .physical_key = input.key.Key.slash, .mods = input.key.Mods{ .shift = false, .ctrl = false, .alt = false, .super = false, .caps_lock = false, .num_lock = false, .sides = input.key.Mods.side{ .shift = input.key.Mods.Side.left, .ctrl = input.key.Mods.Side.left, .alt = input.key.Mods.Side.left, .super = input.key.Mods.Side.left }, ._padding = 0 }, .consumed_mods = input.key.Mods{ .shift = false, .ctrl = false, .alt = false, .super = false, .caps_lock = false, .num_lock = false, .sides = input.key.Mods.side{ .shift = input.key.Mods.Side.left, .ctrl = input.key.Mods.Side.left, .alt = input.key.Mods.Side.left, .super = input.key.Mods.Side.left }, ._padding = 0 }, .composing = false, .utf8 = { 47 }, .unshifted_codepoint = 47 }
debug(surface): created write_req=termio.message.MessageData(u8,38){ .small = termio.message.MessageData(u8,38).Small{ .data = { 47, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170 }, .len = 1 } }
debug(io_thread): mailbox message=termio.message.Message{ .write_small = termio.message.MessageData(u8,38).Small{ .data = { 47, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170, 170 }, .len = 1 } }
```

The event for the Thinkpad keyboard detects the correct code point `47` (which is shown in Ghostty's terminal inspector as well as `kitten show-key -m kitty`), but it detects that the control modifier is being pressed. I'll show the events for both keyboards properly formatted below.

Thinkpad keyboard:

```zig
input.key.KeyEvent{
    .action = input.key.Action.press,
    .key = input.key.Key.slash,
    // Indeed, the physical right control key is replaced by the / key in this keyboard layout
    .physical_key = input.key.Key.right_control,
    .mods = input.key.Mods{
        .shift = false,
        // This is most likely the culprit of the incorrect behavior. The Ctrl
        // key was not being pressed when this event was emitted, so I assume
        // it is inferred somehow. Maybe from the .physical_key?
        .ctrl = true,
        .alt = false,
        .super = false,
        .caps_lock = false,
        .num_lock = false,
        .sides = input.key.Mods.side{
            .shift = input.key.Mods.Side.left,
            .ctrl = input.key.Mods.Side.right,
            .alt = input.key.Mods.Side.left,
            .super = input.key.Mods.Side.left,
        },
        ._padding = 0,
    },
    .consumed_mods = input.key.Mods{
        .shift = false,
        .ctrl = false,
        .alt = false,
        .super = false,
        .caps_lock = false,
        .num_lock = false,
        .sides = input.key.Mods.side{
            .shift = input.key.Mods.Side.left,
            .ctrl = input.key.Mods.Side.left,
            .alt = input.key.Mods.Side.left,
            .super = input.key.Mods.Side.left,
        },
        ._padding = 0,
    },
    .composing = false,
    .utf8 = { 47 },
    .unshifted_codepoint = 47,
};
```

US international keyboard:

```zig
input.key.KeyEvent{
    .action = input.key.Action.press,
    .key = input.key.Key.slash,
    .physical_key = input.key.Key.slash,
    .mods = input.key.Mods{
        .shift = false,
        .ctrl = false,
        .alt = false,
        .super = false,
        .caps_lock = false,
        .num_lock = false,
        .sides = input.key.Mods.side{
            .shift = input.key.Mods.Side.left,
            .ctrl = input.key.Mods.Side.left,
            .alt = input.key.Mods.Side.left,
            .super = input.key.Mods.Side.left,
        },
        ._padding = 0,
    },
    .consumed_mods = input.key.Mods{ .shift = false,
        .ctrl = false,
        .alt = false,
        .super = false,
        .caps_lock = false,
        .num_lock = false,
        .sides = input.key.Mods.side{
            .shift = input.key.Mods.Side.left,
            .ctrl = input.key.Mods.Side.left,
            .alt = input.key.Mods.Side.left,
            .super = input.key.Mods.Side.left,
        },
        ._padding = 0,
    },
    .composing = false,
    .utf8 = { 47 },
    .unshifted_codepoint = 47,
    };
```

The issue at this layer seems to be that Ghostty is setting the `.mods.ctrl` flag on the key event. This means that the bug is at an earlier stage, possibly in how Ghostty generates the `input.key.KeyEvent` struct.

## How key events are generated

This is not very clear to me, but following the usage of the `KeyEvent` struct led me to the `keyCallback` function in `src/Surface.zig`, and later to the `keyEvent` function in `src/apprt/embedded.zig`. As far as I understand the `keyEvent` function is used by exported functions such as `ghostty_surface_key`, which are then exposed to be called by C APIs. As far as I know this is only used by the Swift interface for MacOS (not my use case, but maybe the bug also exists on MacOS). Then I found the `keyEvent` function in `src/apprt/gtk/Surface.zig`, and I suspect that this is where the `KeyEvent` is created, so this is probably the place for the fix.

## Next steps

Investigate how the event is generated in `src/apprt/gtk/Surface.zig` and see if there's anything Ghostty can do to produce the correct behavior. I don't have a Mac so I can't reproduce the bug on another platform, but I think there's a chance that the issue also exists on it (although using a Mac with a thinkpad keyboard seems like a pretty uncommon usecase).

**Comment by luissimas** [comment]
After investigating the way Ghostty generates the `KeyEvent` events on `src/apprt/gtk/Surface.zig`, I found the `eventMods` function in `src/apprt/gtk/key.zig`. The function responsible for getting the modifier state from the device is `gdk_device_get_modifier_state`, and is part of the [GTK API](https://docs.gtk.org/gdk4/method.Device.get_modifier_state.html).

After adding some logs, I verified that this function's return value for pressing the `/` key on the Thinkpad keyboard results in a `GDK_CONTROL_MASK` value (for reference see [the GTK docs](https://docs.gtk.org/gdk4/flags.ModifierType.html)). Pressing `/` or `Ctrl+/` in the Thinkpad keyboard yield the same `GDK_CONTROL_MASK` return value on this function.

https://github.com/ghostty-org/ghostty/blob/94599102e9fb8247af08cbbbcb7ee25e3d31e1bd/src/apprt/gtk/key.zig#L116-L134

I'm not familiar with the GTK APIs, so I don't know if there's a way around this issue. For the moment this seems like a GTK bug and if it is, I don't know if there's much that Ghostty can do in this case. Even doing an ad hoc handling of this specific keyboard layout as is done with some MacOS translations does not seem to be possible (and doesn't sound like a good solution in the first place), due to the fact that detecting the keyboard layout is only supported on MacOS.

I think the reason that this keyboard layout works on Kitty and Alacritty is because they do not source their input events from GTK, so the possible bug does not affect them.

**Reply by mitchellh** [reply, marked as answer]
I've been working on input bugs all morning and was looking at this and came to the same conclusion. GTK is giving us control as being set. I'm not sure if we can do anything about this. If someone has an idea, I'd be very happy to implement it.
